libsql-sqlite3/src/vectorfloat64.c

```diff
@@ -0,0 +1,240 @@
+** FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+** COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+** IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+** CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+**
+******************************************************************************
+**
+** 64-bit floating point vector format utilities.
+*/
+#ifndef SQLITE_OMIT_VECTOR
+#include "sqliteInt.h"
+
+#include "vectorInt.h"
+
+#include <math.h>
+
+/**************************************************************************
+** Utility routines for debugging
+**************************************************************************/
+
+void vectorF64Dump(const Vector *pVec){
+  double *elems = pVec->data;
+  unsigned i;
+  for(i = 0; i < pVec->dims; i++){
+    printf("%lf ", elems[i]);
+  }
+  printf("\n");
+}
+
+/**************************************************************************
+** Utility routines for vector serialization and deserialization
+**************************************************************************/
+
+static inline unsigned formatF64(double value, char *pBuf, int nBufSize){
+  sqlite3_snprintf(nBufSize, pBuf, "%g", value);
+  return strlen(pBuf);
+}
+
+static inline unsigned serializeF64(unsigned char *pBuf, double value){
+  u64 *p = (u64 *)&value;
+  pBuf[0] = *p & 0xFF;
+  pBuf[1] = (*p >> 8) & 0xFF;
+  pBuf[2] = (*p >> 16) & 0xFF;
+  pBuf[3] = (*p >> 24) & 0xFF;
+  pBuf[4] = (*p >> 32) & 0xFF;
+  pBuf[5] = (*p >> 40) & 0xFF;
+  pBuf[6] = (*p >> 48) & 0xFF;
+  pBuf[7] = (*p >> 56) & 0xFF;
+  return sizeof(double);
+}
+
+static inline double deserializeF64(const unsigned char *pBuf){
+  u64 value = 0;
+  value |= (u64)pBuf[0];
+  value |= (u64)pBuf[1] << 8;
+  value |= (u64)pBuf[2] << 16;
+  value |= (u64)pBuf[3] << 24;
+  value |= (u64)pBuf[4] << 32;
+  value |= (u64)pBuf[5] << 40;
+  value |= (u64)pBuf[6] << 48;
+  value |= (u64)pBuf[7] << 56;
+  return *(double *)&value;
+}
+
+size_t vectorF64SerializeToBlob(
+  const Vector *pVector,
+  unsigned char *pBlob,
+  size_t nBlobSize
+){
+  double *elems = pVector->data;
+  unsigned char *pPtr = pBlob;
+  size_t len = 0;
+  unsigned i;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT64 );
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= pVector->dims * sizeof(double) );
+
+  for (i = 0; i < pVector->dims; i++) {
+    pPtr += serializeF64(pPtr, elems[i]);
+    len += sizeof(double);
+  }
+  return len;
+}
+
+size_t vectorF64DeserializeFromBlob(
+  Vector *pVector,
+  const unsigned char *pBlob,
+  size_t nBlobSize
+){
+  double *elems = pVector->data;
+  unsigned i;
+  pVector->type = VECTOR_TYPE_FLOAT64;
+  pVector->dims = nBlobSize / sizeof(double);
+
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= vectorDataSize(pVector->type, pVector->dims) );
```

Comment:
Seems redundant because pVector->dims is set to nBlobSize / sizeof(double) 3 lines above.